### DISCUSSION TITLE
Applying Neumann BCs to Randomly Distributed Spherical Particles

### DISCUSSION THREAD

**Discussion by Josh-Web-700** [idaholab/moose#28663] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello,

I am working on a phase-field solidification simulation. I will be trying to create a set of randomly distributed solid spheres in the domain, each with a no-flux boundary condition applied at the surface.

I was wondering if it would be possible to use MultiSmoothCircleIC for creating an initial condition with multiple spherical solid particles in the domain, and then to apply a no flux Neumann boundary condition to the surface of each of these particles (the dot product of the gradient of the order parameter with the sphere's surface normals should be zero).

What is the best way to apply this BC? Is there any built-in functionality in MOOSE for this? If not, I was wondering if it is possible to modify MultiSmoothCircleIC so that it exports the locations and sizes of each of the spherical particles to a set of variables that can be read by the BCs or Mesh system. Then, the BCs system would somehow need to be able to compute the normal to the nodes on the surface of the sphere and apply the boundary condition.

If I understand the MOOSE system and my problem correctly, I have three questions:

1. Would it be better to supply a mesh with randomly distributed spherical "holes," and apply the BCs to these internal surfaces, or would it be better to begin with a uniform rectangular mesh and apply the BCs to some defined spherical surfaces contained in the domain?
2. If I do need to supply a mesh with randomly distributed spherical holes, how would I customize the BCs system in MOOSE to apply a Neumann BC to these interior surfaces?
3. If it is possible to create these BCs on a uniform rectangular mesh _without any "holes"_, what is the best method and syntax for doing so?

From my understanding, it looks like I might be able to use ParsedGenerateSideset to define spherical sidesets to apply these BCs, but I am just not sure if this is the most efficient way to do so.

Please let me know if you need further clarification. If this is possible with MOOSE, it would be greatly appreciated if you could guide me to the best way to do this. Thank you so much!

**Comment by GiudGiud** [comment, marked as answer]
Hello

1) would certainly be good. You can do this in gmsh, in cubit learn, or use the XYZDelaunayTriangulator which just got added yesterday (in the next branch only right now) with SphereMeshGenerators for the spheres (and transformGenerator to translate the spheres) 

2) BCs are not applied inside the domain of a variable. Is the variable you plan to apply the BC on defined on both sides of the spherical boundaries?

3) MeshGenerators (such as the ParsedGenerateSideset), which are in charge of modifying the mesh at initialization, cannot read variable values, let alone gradient. So instead you will want to create a MeshModifier, and look at the variable gradient on every element side.

**Reply by Josh-Web-700** [reply]
Just to clarify, what is the purpose of using a MeshModifier? Would that be for a dynamic or changing mesh, or is there another reason to use a MeshModifier for a mesh like mine that would be static and unchanging with time?

Thank you again!

**Reply by GiudGiud** [reply]
The intent of the MeshModifier system is to modify the mesh. It can be for dynamically modifying the mesh, or it can be executed on the INITIAL execue_on flag, only once. The difference with mesh generators is that mesh modifiers can couple to variables, because they execute after the setup phase

**Reply by Josh-Web-700** [reply]
That makes sense, thank you. That way, the Mesh Modifier can modify the mesh to adapt to the MultiSmoothCircleIC condition, and do so only on the INITIAL execute_on flag.

Thank you again for your help!

**Comment by Josh-Web-700** [comment]
Yes, in case (2) my variable would be defined on both sides of the spherical boundaries, and I see now why that wouldn't work to apply a boundary condition. That is helpful, thank you so much!
